src/libarena/lib.rs

```diff
@@ -13,6 +13,7 @@
 
 #![deny(rust_2018_idioms)]
 #![deny(internal)]
+#![deny(unused_lifetimes)]
```

Comment:
Both `deny(internal)` and `deny(unused_lifetimes)` can be enabled through rustbuild (`src\bootstrap\bin\rustc.rs`) rather than through all crates individually, similarly to `deny(warnings)`/`deny(bare_trait_objects)`.

Reply:
Fair point. cc @flip1995

Reply:
Did this locally and it revealed `rustc_data_structures` and `rustc_macros` were missing `deny(internal)` and `deny(unused_lifetimes)`.

Reply:
Looks like `default_hash_types` may be a problem.
Unlike other lints it's not universally applicable and is not a no-op for non-compiler crates.

Reply:
It's even worse, the `rustbuild` thing seems to apply to all crates, not just compiler ones.

Reply:
> the rustbuild thing seems to apply to all crates

I expected that, but also expected `internal` lints to not report anything for non-compiler crates (since they don't have `TyCtxt` etc).

Reply:
Yeah `default_hash_types` is also explicitly allowed in `libstd` for example. 

Most of the internal lints shouldn't report anything in most of the crates. I will figure this out in #61545

Reply:
Any reason why these are `deny` instead of `warn`? This means these are hard errors even with `deny-warnings = false`.